src/Razor/Microsoft.AspNetCore.Razor.Language/test/IntegrationTests/ComponentCodeGenerationTestBase.cs

```diff
@@ -3152,6 +3152,391 @@ public class MyComponent<TItem> : ComponentBase
+    }
+}
+"));
+
+            // Act
+            var generated = CompileToCSharp(@"
+<Grid Items=""@(Array.Empty<DateTime>())""><Column>@context.Year</Column></Grid>");
+
+            // Assert
+            AssertDocumentNodeMatchesBaseline(generated.CodeDocument);
+            AssertCSharpDocumentMatchesBaseline(generated.CodeDocument);
+            CompileToAssembly(generated);
+        }
+
+        [Fact]
+        public void ChildComponent_Generic_TypeInference_CascadedWithLambda()
+        {
+            // Arrange
+            AdditionalSyntaxTrees.Add(Parse(@"
+using Microsoft.AspNetCore.Components;
+
+namespace Test
+{
+    [CascadingTypeParameter(nameof(TItem))]
+    public class Grid<TItem> : ComponentBase
+    {
+        [Parameter] public System.Collections.Generic.IEnumerable<TItem> Items { get; set; }
+        [Parameter] public RenderFragment ChildContent { get; set; }
+    }
+
+    public class Column<TItem, TOutput> : ComponentBase
+    {
+        [Parameter] public System.Func<TItem, TOutput> SomeLambda { get; set; }
+    }
+}
+"));
+
+            // Act
+            var generated = CompileToCSharp(@"
+<Grid Items=""@(Array.Empty<DateTime>())""><Column SomeLambda=""@(x => x.Year)"" /></Grid>");
+
+            // Assert
+            AssertDocumentNodeMatchesBaseline(generated.CodeDocument);
+            AssertCSharpDocumentMatchesBaseline(generated.CodeDocument);
+            CompileToAssembly(generated);
+        }
+
+        [Fact]
+        public void ChildComponent_Generic_TypeInference_CascadedWithMultipleTypes()
+        {
+
+            // Arrange
+            AdditionalSyntaxTrees.Add(Parse(@"
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Components;
+
+namespace Test
+{
+    [CascadingTypeParameter(nameof(TKey))]
+    [CascadingTypeParameter(nameof(TValue))]
+    [CascadingTypeParameter(nameof(TOther))]
+    public class Parent<TKey, TValue, TOther> : ComponentBase
+    {
+        [Parameter] public Dictionary<TKey, TValue> Data { get; set; }
+        [Parameter] public TOther Other { get; set; }
+        [Parameter] public RenderFragment ChildContent { get; set; }
+    }
+
+    public class Child<TOther, TValue, TKey, TChildOnly> : ComponentBase
+    {
+        [Parameter] public ICollection<TChildOnly> ChildOnlyItems { get; set; }
+    }
+}
+"));
+
+            // Act
+            var generated = CompileToCSharp(@"
+<Parent Data=""@(new System.Collections.Generic.Dictionary<int, string>())"" Other=""@DateTime.MinValue"">
+    <Child ChildOnlyItems=""@(new[] { 'a', 'b', 'c' })"" />
+</Parent>");
+
+            // Assert
+            AssertDocumentNodeMatchesBaseline(generated.CodeDocument);
+            AssertCSharpDocumentMatchesBaseline(generated.CodeDocument);
+            CompileToAssembly(generated);
+        }
+
+        [Fact]
+        public void ChildComponent_Generic_TypeInference_CascadedWithUnrelatedGenericType_CreatesDiagnostic()
```

Comment:
We might want to capture some other behaviors here just to make sure we don't unknowingly change them in the future. I believe this test covers the case where the parent cascades more values than the child. There are a few other cases that might be interesting to cover:
* Same number of parameters but different names.
* Child has more generic parameters than the parent.
* Combined ancestors provide all generic parameters for a child component but none of them covers all of them.
* Child has more generic parameters, the parent covers some of them and the rest are explicitly specified.

We might already have some of these tests, but I'm not sure I've seen them covered.

Reply:
> Same number of parameters but different names.

The non-matchingness of different names is covered by `ChildComponent_Generic_TypeInference_CascadedWithMultipleTypes`. I don't think we specifically need a test about the number of parameters being the same, as that's not a consequential aspect in either the design or implementation.

> Child has more generic parameters than the parent.
> Child has more generic parameters, the parent covers some of them and the rest are explicitly specified.

These are covered by `ChildComponent_Generic_TypeInference_CascadedWithMultipleTypes`, at least in the case where the extra child parameter is inferred. We could add another one for the extra child parameter being explicit but there's really no difference in terms of the implementation for that so it seems redundant.

> Combined ancestors provide all generic parameters for a child component but none of them covers all of them.

Good idea. Added.